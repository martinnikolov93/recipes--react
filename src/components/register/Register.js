import React from 'react';

import { Link } from 'react-router-dom'

import { isEmptyObject } from '../../utils/helpers'
import UserContext from '../../Context';

class Register extends React.Component {
    state = {
        email: '',
        emailErrors: false,
        password: '',
        passwordErrors: false,
        rePassword: '',
        rePasswordErrors: false
    }

    passwordValidator = (password) => {
        let passwordErrors = {}

        if (!password) {
            passwordErrors.required = 'Password is required.';
        }

        if (password.length < 5) {
            passwordErrors.length = 'Password must be at least 5 characters.';
        }

        if (isEmptyObject(passwordErrors)) {
            return false
        }

        return passwordErrors
    }

    emailValidator = (email) => {
        let emailErrors = {}

        if (!email) {
            emailErrors.required = 'Email is required.';
        }

        if (isEmptyObject(emailErrors)) {
            return false
        }

        return emailErrors
    }

    rePasswordValidator = (rePassword) => {
        const password = this.state.password

        let rePasswordErrors = {}

        if (!rePassword) {
            rePasswordErrors.required = 'Re-Password is required.';
        }

        if (password !== rePassword) {
            rePasswordErrors.doesNotMatch = 'Passwords does not match.';
        }

        if (isEmptyObject(rePasswordErrors)) {
            return false
        }

        return rePasswordErrors
    }

    submitHandler = (event) => {
        event.preventDefault()

        let { email, password, rePassword } = this.state

        const emailErrors = this.emailValidator(email)
        const passwordErrors = this.passwordValidator(password)
        const rePasswordErrors = this.rePasswordValidator(rePassword)

        this.setState({ emailErrors, passwordErrors, rePasswordErrors })

        if (emailErrors || passwordErrors || rePasswordErrors) {
            return
        }

        fetch('http://localhost:9999/api/user/register', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                email,
                password
            })
        })
            .then((response) => {
                const authToken = response.headers.get('auth-token')
                document.cookie = `auth-token=${authToken}`
                return response.json()
            })
            .then((user) => {
                this.context.logIn(user)
            })
    }

    changeHandlerEmail = (event) => {
        const email = event.target.value
        const emailErrors = this.emailValidator(email)
        this.setState({ email: event.target.value, emailErrors: emailErrors })
    }

    changeHandlerPassword = (event) => {
        const password = event.target.value
        const passwordErrors = this.passwordValidator(password)
        this.setState({ password: password, passwordErrors: passwordErrors })
    }


    changeHandlerRePassword = (event) => {
        const rePassword = event.target.value
        const rePasswordErrors = this.rePasswordValidator(rePassword)
        this.setState({ rePassword: rePassword, rePasswordErrors: rePasswordErrors })
    }

    render() {
        let { email, emailErrors, password, passwordErrors, rePassword, rePasswordErrors } = this.state

        return (
            <div>
                <form onSubmit={this.submitHandler}>
                    <div>
                        <label htmlFor='email'></label>
                        <input className={emailErrors ? "input-error" : ""} id='email' type="email" placeholder="Email" value={email} onChange={this.changeHandlerEmail} />
                        <div className="input-error-text">{emailErrors ? Object.values(emailErrors)[0] : null}</div>
                    </div>
                    <div>
                        <label htmlFor='password'></label>
                        <input className={passwordErrors ? "input-error" : ""} id='password' type="password" placeholder="Password" value={password} onChange={this.changeHandlerPassword} />
                        <div className="input-error-text">{passwordErrors ? Object.values(passwordErrors)[0] : null}</div>
                    </div>
                    <div>
                        <label htmlFor='rePassword'></label>
                        <input className={rePasswordErrors ? "input-error" : ""} id='rePassword' type="password" placeholder="Repeat Password" value={rePassword} onChange={this.changeHandlerRePassword} />
                        <div className="input-error-text">{rePasswordErrors ? Object.values(rePasswordErrors)[0] : null}</div>
                    </div>
                    <button>Register</button>
                </form>
                <br />
                <div>Already have an account? Go to <Link to='/login'>Login Page</Link></div>
            </div>
        )
    }
}

Register.contextType = UserContext;

export default Register