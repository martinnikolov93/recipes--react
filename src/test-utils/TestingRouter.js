import React from 'react';
import UserContext from '../Context';
import { BrowserRouter, Route } from 'react-router-dom'

const TestingRouter = ({user, children}) => {
    return (
        <BrowserRouter>
            <Route>
                <UserContext.Provider value={user}>
                    {children}
                </UserContext.Provider>
            </Route>
        </BrowserRouter>
    )
}

export default TestingRouter