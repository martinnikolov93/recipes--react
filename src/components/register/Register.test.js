import React from 'react';
import renderer from 'react-test-renderer'
import Register from './Register';
import TestingRouter from '../../test-utils/TestingRouter';

describe('Register component', () => {
    it('Should render Register component', () => {
        const tree = renderer.create(
            <TestingRouter>
                <Register />
            </TestingRouter>
        ).toJSON()

        expect(tree).toMatchSnapshot()
    })
})