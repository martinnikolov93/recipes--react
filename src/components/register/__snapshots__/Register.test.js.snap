// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register component Should render Register component 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label
        htmlFor="email"
      />
      <input
        className=""
        id="email"
        onChange={[Function]}
        placeholder="Email"
        type="email"
        value=""
      />
      <div
        className="input-error-text"
      />
    </div>
    <div>
      <label
        htmlFor="password"
      />
      <input
        className=""
        id="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <div
        className="input-error-text"
      />
    </div>
    <div>
      <label
        htmlFor="rePassword"
      />
      <input
        className=""
        id="rePassword"
        onChange={[Function]}
        placeholder="Repeat Password"
        type="password"
        value=""
      />
      <div
        className="input-error-text"
      />
    </div>
    <button>
      Register
    </button>
  </form>
  <br />
  <div>
    Already have an account? Go to 
    <a
      href="/login"
      onClick={[Function]}
    >
      Login Page
    </a>
  </div>
</div>
`;
